### DISCUSSION TITLE
CFG_TUD_MSC_EP_BUFSIZE greater than 64K

### DISCUSSION THREAD

**Discussion by mauriciocirelli** [hathach/tinyusb#1638] [Q&A]
Dear,

Is there any reason to assert CFG_TUD_MSC_EP_BUFSIZE < UIN16_MAX?
I would like to increase it to 128K as I have flash memories with 128K sectors.

Thank you!

**Comment by hathach** [comment]
since most low level API capped at 16-bit length for an call, going larger does not gain much in term of throughput. `CFG_TUD_MSC_EP_BUFSIZE` is only for holding data transferring via USB bus, it does not really need to match your block size. You can use tud_msc_write10_cb/tud_msc_read10_cb as normally does, tinyusb will take care of breaking the 128KB sector into multiple `CFG_TUD_MSC_EP_BUFSIZE` chunk. For portability, you should still define block size = 512.

btw which mcu you are using ?

**Comment by mauriciocirelli** [comment]
Hi,

I am using STM32H7. The problem is the memory part.. The minimum erasable block size is 128K, which makes writing a buffer a more complicated task, since we have to read the block, update it with the buffer in RAM and, then, erase the block and program the new block data. When we work with small chunks, we need to repeat it many times, which makes the process slower and also demands lots of erase operations, reducing the memory endurance. We are using 32K chunks and it is working, but it requires 4 read/erase/program cycles to write a whole 128K block of data. Using greater chunks would make this process a bit faster and would improve the memory endurance.

But I understand it may not be possible. It is something we will have to live with.

**Comment by hathach** [comment]
I have also done some flash caching like this for other mcu, maybe this would help you out https://github.com/adafruit/Adafruit_nRF52_Arduino/blob/master/libraries/InternalFileSytem/src/flash/flash_cache.c

**Comment by mauriciocirelli** [comment]
It is interesting, thanks for sharing.
It may be useful in this case.
